component/tests/common.py

```diff
@@ -0,0 +1,264 @@
+# Copyright 2017 Camptocamp SA
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html)
+
+import copy
+import sys
+import unittest
+from contextlib import contextmanager
+
+import odoo
+from odoo import api
+from odoo.tests import common
+
+from odoo.addons.component.core import ComponentRegistry, MetaComponent, _get_addon_name
+
+
+@contextmanager
+def new_rollbacked_env():
+    registry = odoo.registry(common.get_db_name())
+    uid = odoo.SUPERUSER_ID
+    cr = registry.cursor()
+    try:
+        yield api.Environment(cr, uid, {})
+    finally:
+        cr.rollback()  # we shouldn't have to commit anything
+        cr.close()
+
+
+class ComponentMixin(object):
+    @classmethod
+    def setUpComponent(cls):
+        with new_rollbacked_env() as env:
+            builder = env["component.builder"]
+            # build the components of every installed addons
+            comp_registry = builder._init_global_registry()
+            cls._components_registry = comp_registry
+            # ensure that we load only the components of the 'installed'
+            # modules, not 'to install', which means we load only the
+            # dependencies of the tested addons, not the siblings or
+            # chilren addons
+            builder.build_registry(comp_registry, states=("installed",))
+            # build the components of the current tested addon
+            current_addon = _get_addon_name(cls.__module__)
+            env["component.builder"].load_components(current_addon)
+            if hasattr(cls, "env"):
+                cls.env.context = dict(
+                    cls.env.context, components_registry=cls._components_registry
+                )
+
+    # pylint: disable=W8106
+    def setUp(self):
+        # should be ready only during tests, never during installation
+        # of addons
+        self._components_registry.ready = True
+
+        @self.addCleanup
+        def notready():
+            self._components_registry.ready = False
+
+
+class TransactionComponentCase(common.TransactionCase, ComponentMixin):
+    """A TransactionCase that loads all the components
+
+    It it used like an usual Odoo's TransactionCase, but it ensures
+    that all the components of the current addon and its dependencies
+    are loaded.
+
+    """
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.setUpComponent()
+
+    # pylint: disable=W8106
+    def setUp(self):
+        # resolve an inheritance issue (common.TransactionCase does not call
+        # super)
+        common.TransactionCase.setUp(self)
+        ComponentMixin.setUp(self)
+        # There's no env on setUpClass of TransactionCase, must do it here.
+        self.env.context = dict(
+            self.env.context, components_registry=self._components_registry
+        )
+
+
+class SavepointComponentCase(common.SavepointCase, ComponentMixin):
```

Comment:
This class should be deprecated as common.SavepointCase.
I would make it inherit from `TransactionComponentCase` directly and just log deprecation.
Setup code should be moved from setUp to setUpClass.